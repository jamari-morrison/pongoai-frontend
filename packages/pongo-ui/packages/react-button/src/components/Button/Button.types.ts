import type { ComponentProps, ComponentState, IntrinsicShorthandProps } from '@fluentui/react-utilities';
import { ARIAButtonShorthandProps } from '@fluentui/react-aria';
export type ButtonSlots = {
  /**
   * The root of the Button.
   */
  //root: IntrinsicShorthandProps<'button'> | IntrinsicShorthandProps<'a'>;
  root: ARIAButtonShorthandProps;
};

export type ButtonCommons = {
  /**
   * The appearance variant to use for the button.
   *
   * @default outline
   */
  appearance?: 'outline' | 'primary' | 'subtle' | 'transparent';

  /**
   * Whether the button should be disabled.
   *
   * @default false
   */
  disabled?: boolean;

  /**
   * The shape of the button.
   *
   * @default rounded
   */
  shape?: 'rounded' | 'circular' | 'square';
};

export type ButtonProps = ComponentProps<ButtonSlots> & ButtonCommons;

export type ButtonState = ComponentState<ButtonSlots> & ButtonCommons;
