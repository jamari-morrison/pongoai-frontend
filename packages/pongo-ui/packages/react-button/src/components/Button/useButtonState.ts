import { useEventCallback } from '@fluentui/react-utilities';
import type { ButtonState } from './Button.types';

export const useButtonState = (state: ButtonState): ButtonState => {
  const { disabled } = state;
  const {
    as,
    tabIndex,
    role,
    onClick: onClickCallback,
    onKeyDown: onKeyDownCallback,
    onKeyUp: onKeyUpCallback,
  } = state.root;

  /**
   * Button event callback for when the root is not rendered as a button (EX: anchor).
   */
  const onButtonClick = useEventCallback((ev: any) => {
    if (!disabled && !ev.isDefaultPrevented()) {
      onClickCallback?.(ev);
    } else {
      ev.preventDefault();
      ev.stopPropagation();
    }
  });

  /**
   * Button event callback for when the root is not rendered as a button (EX: anchor).
   */
  const onButtonKeydown = useEventCallback((ev: any) => {
    if (!disabled && !ev.isDefaultPrevented()) {
      onKeyDownCallback?.(ev);
      switch (ev.key) {
        case 'Enter':
          ev.preventDefault();
          ev.currentTarget.click();
          break;
        case 'Space':
          ev.preventDefault();
          break;
        default:
      }
    } else {
      ev.preventDefault();
      ev.stopPropagation();
    }
  });

  /**
   * Button event callback for when the root is not rendered as a button (EX: anchor).
   */
  const onButtonKeyup = useEventCallback((ev: any) => {
    if (!disabled && !ev.isDefaultPrevented()) {
      onKeyUpCallback?.(ev);
      switch (ev.key) {
        case 'Enter':
          ev.preventDefault();
          ev.currentTarget.click();
          break;
        case 'Space':
          ev.preventDefault();
          break;
        default:
      }
    } else {
      ev.preventDefault();
      ev.stopPropagation();
    }
  });

  // If the button is rendered as an anchor or other element apply these props:
  if (as !== 'button') {
    state.root['aria-disabled'] = disabled;
    state.root.onClick = onButtonClick;
    state.root.onKeyDown = onButtonKeydown;
    state.root.onKeyUp = onButtonKeyup;
    state.root.role = role ?? 'button';
    state.root.tabIndex = disabled ? undefined : tabIndex ?? 0;
  }
  // Otherwise if it is a button:
  else {
    state.root.disabled = disabled;
  }

  return state;
};
