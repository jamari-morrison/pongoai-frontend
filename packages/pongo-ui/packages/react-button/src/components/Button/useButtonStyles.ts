import { makeStyles, mergeClasses } from '@fluentui/react-make-styles';
import { createFocusIndicatorStyleRule } from '@fluentui/react-tabster';
import type { ButtonState } from './Button.types';
import { Theme } from '@pongoai/react-theme';

const useRootStyles = makeStyles({
  root: (theme: Theme) => ({
    position: 'relative',
    display: 'inline-flex',
    alignItems: 'center',
    justifyContent: 'center',
    verticalAlign: 'middle',
    transition: 'background 100ms cubic-bezier(0.4, 0, 0.2, 1) 0ms, box-shadow 100ms cubic-bezier(0.4, 0, 0.2, 1) 0ms',
    fontFamily: theme.fonts.fontFamily.base,
    fontStyle: 'normal',
    fontWeight: '500',
    outline: 'none',
  }),

  focusIndicator: createFocusIndicatorStyleRule(
    {
      ':after': {
        content: "''",
        position: 'absolute',
        top: '-8px',
        right: '-8px',
        bottom: '-8px',
        left: '-8px',
        boxSizing: 'border-box',
        border: `2px solid black`,
        borderRadius: '4px',
      },
    },
    { selector: 'focus-within' },
  ),

  /**
   * Disabled/Enabled styles
   */
  enabled: {
    cursor: 'pointer',
    '&:hover': {
      boxShadow:
        '0px 3px 1px -2px rgb(0 0 0 / 20%), 0px 2px 2px 0px rgb(0 0 0 / 14%), 0px 1px 5px 0px rgb(0 0 0 / 12%)',
    },
    '&:active': {
      boxShadow: 'none',
    },
  },

  disabled: {
    cursor: 'not-allowed',
  },

  /**
   * Size styles
   */
  small: (theme: Theme) => ({
    minWidth: '64px',
    height: '30px',
    fontSize: theme.fonts.fontSize[200],
  }),

  medium: (theme: Theme) => ({
    minWidth: '80px',
    height: '35px',
    fontSize: theme.fonts.fontSize[300],
  }),

  large: (theme: Theme) => ({
    minWidth: '96px',
    height: '45px',
    fontSize: theme.fonts.fontSize[400],
  }),

  /**
   * Color styles enabled
   */
  inherit: (theme: Theme) => ({
    color: theme.palette.inherit,
    border: `2px solid ${theme.palette.inherit}`,
    background: theme.palette.neutral,
    backgroundColor: theme.palette.inherit,

    '&:hover': {
      background: theme.palette.inheritForegroundHover,
      borderColor: theme.palette.inheritHover,
      color: theme.palette.inheritHover,
      backgroundColor: theme.palette.inheritHover,
    },
    '&:active': {
      background: theme.palette.inheritForegroundPressed,
      borderColor: theme.palette.inheritPressed,
      color: theme.palette.inheritPressed,
      backgroundColor: theme.palette.inheritPressed,
    },
  }),

  inheritDisabled: (theme: Theme) => ({
    color: theme.palette.inheritDisabled,
    border: `2px solid ${theme.palette.inheritDisabled}`,
    backgroundColor: theme.palette.inheritDisabled,
    background: theme.palette.neutral,
  }),

  brand: (theme: Theme) => ({
    color: theme.palette.brand,
    border: `2px solid ${theme.palette.brand}`,
    background: theme.palette.neutral,
    backgroundColor: theme.palette.brand,

    '&:hover': {
      background: theme.palette.brandForegroundHover,
      borderColor: theme.palette.brandHover,
      color: theme.palette.brandHover,
      backgroundColor: theme.palette.brandHover,
    },
    '&:active': {
      background: theme.palette.brandForegroundPressed,
      borderColor: theme.palette.brandPressed,
      color: theme.palette.brandPressed,
      backgroundColor: theme.palette.brandPressed,
    },
  }),

  brandDisabled: (theme: Theme) => ({
    color: theme.palette.brandDisabled,
    border: `2px solid ${theme.palette.brandDisabled}`,
    backgroundColor: theme.palette.brandDisabled,
    background: theme.palette.neutral,
  }),

  secondary: (theme: Theme) => ({
    color: theme.palette.secondary,
    border: `2px solid ${theme.palette.secondary}`,
    background: theme.palette.neutral,
    backgroundColor: theme.palette.secondary,

    '&:hover': {
      background: theme.palette.secondaryForegroundHover,
      borderColor: theme.palette.secondaryHover,
      color: theme.palette.secondaryHover,
      backgroundColor: theme.palette.secondaryHover,
    },
    '&:active': {
      background: theme.palette.secondaryForegroundPressed,
      borderColor: theme.palette.secondaryPressed,
      color: theme.palette.secondaryPressed,
      backgroundColor: theme.palette.secondaryPressed,
    },
  }),

  secondaryDisabled: (theme: Theme) => ({
    color: theme.palette.secondaryDisabled,
    border: `2px solid ${theme.palette.secondaryDisabled}`,
    backgroundColor: theme.palette.secondaryDisabled,
    background: theme.palette.neutral,
  }),

  success: (theme: Theme) => ({
    color: theme.palette.success,
    border: `2px solid ${theme.palette.success}`,
    background: theme.palette.neutral,
    backgroundColor: theme.palette.success,

    '&:hover': {
      background: theme.palette.successForegroundHover,
      borderColor: theme.palette.successHover,
      color: theme.palette.successHover,
      backgroundColor: theme.palette.successHover,
    },
    '&:active': {
      background: theme.palette.successForegroundPressed,
      borderColor: theme.palette.successPressed,
      color: theme.palette.successPressed,
      backgroundColor: theme.palette.successPressed,
    },
  }),

  successDisabled: (theme: Theme) => ({
    color: theme.palette.successDisabled,
    border: `2px solid ${theme.palette.successDisabled}`,
    backgroundColor: theme.palette.successDisabled,
    background: theme.palette.neutral,
  }),

  error: (theme: Theme) => ({
    color: theme.palette.error,
    border: `2px solid ${theme.palette.error}`,
    background: theme.palette.neutral,
    backgroundColor: theme.palette.error,

    '&:hover': {
      background: theme.palette.errorForegroundHover,
      borderColor: theme.palette.errorHover,
      color: theme.palette.errorHover,
      backgroundColor: theme.palette.errorHover,
    },
    '&:active': {
      background: theme.palette.errorForegroundPressed,
      borderColor: theme.palette.errorPressed,
      color: theme.palette.errorPressed,
      backgroundColor: theme.palette.errorPressed,
    },
  }),

  errorDisabled: (theme: Theme) => ({
    color: theme.palette.errorDisabled,
    border: `2px solid ${theme.palette.errorDisabled}`,
    backgroundColor: theme.palette.errorDisabled,
    background: theme.palette.neutral,
  }),

  social: (theme: Theme) => ({
    color: theme.palette.social,
    border: `2px solid ${theme.palette.social}`,
    background: theme.palette.neutral,
    backgroundColor: theme.palette.social,

    '&:hover': {
      background: theme.palette.socialForegroundHover,
      borderColor: theme.palette.socialHover,
      color: theme.palette.socialHover,
      backgroundColor: theme.palette.socialHover,
    },
    '&:active': {
      background: theme.palette.socialForegroundPressed,
      borderColor: theme.palette.socialPressed,
      color: theme.palette.socialPressed,
      backgroundColor: theme.palette.socialPressed,
    },
  }),

  socialDisabled: (theme: Theme) => ({
    color: theme.palette.socialDisabled,
    border: `2px solid ${theme.palette.socialDisabled}`,
    backgroundColor: theme.palette.socialDisabled,
    background: theme.palette.neutral,
  }),

  warning: (theme: Theme) => ({
    color: theme.palette.warning,
    border: `2px solid ${theme.palette.warning}`,
    background: theme.palette.neutral,
    backgroundColor: theme.palette.warning,

    '&:hover': {
      background: theme.palette.warningForegroundHover,
      borderColor: theme.palette.warningHover,
      color: theme.palette.warningHover,
      backgroundColor: theme.palette.warningHover,
    },
    '&:active': {
      background: theme.palette.warningForegroundPressed,
      borderColor: theme.palette.warningPressed,
      color: theme.palette.warningPressed,
      backgroundColor: theme.palette.warningPressed,
    },
  }),

  warningDisabled: (theme: Theme) => ({
    color: theme.palette.warningDisabled,
    border: `2px solid ${theme.palette.warningDisabled}`,
    backgroundColor: theme.palette.warningDisabled,
    background: theme.palette.neutral,
  }),

  info: (theme: Theme) => ({
    color: theme.palette.info,
    border: `2px solid ${theme.palette.info}`,
    background: theme.palette.neutral,
    backgroundColor: theme.palette.info,

    '&:hover': {
      background: theme.palette.infoForegroundHover,
      borderColor: theme.palette.infoHover,
      color: theme.palette.infoHover,
      backgroundColor: theme.palette.infoHover,
    },
    '&:active': {
      background: theme.palette.infoForegroundPressed,
      borderColor: theme.palette.infoPressed,
      color: theme.palette.infoPressed,
      backgroundColor: theme.palette.infoPressed,
    },
  }),

  infoDisabled: (theme: Theme) => ({
    color: theme.palette.infoDisabled,
    border: `2px solid ${theme.palette.infoDisabled}`,
    backgroundColor: theme.palette.infoDisabled,
    background: theme.palette.neutral,
  }),

  /**
   * Appearance styles
   */
  outline: {
    backgroundColor: 'none !important',

    '&:hover': {
      backgroundColor: 'none',
    },
    '&:active': {
      backgroundColor: 'none',
    },
  },

  primary: (theme: Theme) => ({
    border: 'none',
    color: 'white !important',
  }),

  subtle: {
    border: 'none',
    backgroundColor: 'none',

    '&:hover': {
      backgroundColor: 'none',
    },
    '&:active': {
      backgroundColor: 'none',
    },
  },

  transparent: {
    border: 'none !important',
    background: 'none !important',
  },

  transparentEnabled: {
    '&:hover': {
      boxShadow: 'none',
      textDecoration: 'underline',
    },
    '&:active': {
      textDecoration: 'underline',
    },
  },

  /**
   * Shape styles
   */
  rounded: {
    borderRadius: '6px',
  },

  circular: {
    borderRadius: '999px',
  },

  circle: {
    borderRadius: '100%',
  },

  square: {
    borderRadius: '0px',
  },
});

export const useButtonStyles = (state: ButtonState) => {
  const rootStyles = useRootStyles();

  state.root.className = mergeClasses(
    rootStyles.root,
    state.disabled ? (rootStyles as any)[state.color! + 'Disabled'] : rootStyles[state.color!],
    !state.disabled && rootStyles.focusIndicator,
    state.disabled ? rootStyles.disabled : rootStyles.enabled,
    rootStyles[state.appearance!],
    rootStyles[state.size!],
    rootStyles[state.shape!],
    state.appearance === 'transparent' && !state.disabled && rootStyles.transparentEnabled,
    state.root.className,
  );

  return state;
};
