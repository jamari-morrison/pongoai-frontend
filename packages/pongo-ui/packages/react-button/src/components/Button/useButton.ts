import * as React from 'react';
import { getNativeElementProps } from '@fluentui/react-utilities';
import { useButtonState } from './useButtonState';
import type { ButtonProps, ButtonSlots, ButtonState } from './Button.types';

export const buttonShorthandProps: (keyof ButtonSlots)[] = ['root'];

export const useButton = (props: ButtonProps, ref: React.Ref<HTMLElement>): ButtonState => {
  const { as, appearance = 'outline', disabled = false, shape = 'rounded' } = props;

  const state: ButtonState = {
    appearance,
    disabled,
    shape,
    components: {
      root: 'button',
    },
    root: getNativeElementProps(as || 'button', {
      ref,
      required: true,
      type: 'button',
      ...props,
    }),
  };

  useButtonState(state);

  return state;
};
