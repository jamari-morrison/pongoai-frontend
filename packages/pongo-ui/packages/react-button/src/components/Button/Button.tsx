import * as React from 'react';
import { useButton } from './useButton';
import { renderButton } from './renderButton';
import { useButtonStyles } from './useButtonStyles';
import type { ButtonProps } from './Button.types';

/**
 * The Button allows a user to make an action through a press.
 */
export const Button = React.forwardRef<HTMLDivElement, ButtonProps>((props, ref) => {
  const state = useButton(props, ref);

  useButtonStyles(state);

  return renderButton(state);
});
Button.displayName = 'Button';
