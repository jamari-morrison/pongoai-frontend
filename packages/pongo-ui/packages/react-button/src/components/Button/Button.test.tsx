import { Button } from './Button';
import { isConformant } from '../../common/isConformant';

describe('Button', () => {
  isConformant({
    Component: Button,
    displayName: 'Button',
  });
});
