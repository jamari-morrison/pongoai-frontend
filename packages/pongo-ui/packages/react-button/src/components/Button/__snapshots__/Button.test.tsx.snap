// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button when rendered as an anchor renders correctly 1`] = `
<div>
  <a
    aria-disabled="false"
    class="___11a081q f10pi13n ftuwxu6 f122n59 f4d9j23 fmrv4ls f14z66ap fcv2ht4 fmbxnhj f1e4rdhr fmd4ok8 fcsweqw f6kqwle f1s6fcnf f2hkw1w fdxsi5y f1i4q40k f1ab1e41 f1jd910e f1nw9czy f21wkpv fvk4ntw f1bkt4b4 f16hf5f2 fyyqeim f1qhqcal f1kbdjx9 fw1d893 f14hlsw1 fpuz8dn f1pyiwo4 f19m6738 f1ioparz ftwtw2f f184ipub fcobno7 f1o9ynst f4blu93 f1k6fduh f1y5ik6t f1d3e705 ftac7j7 f1w0yd9v fdwcyh7 f1h0xgbp fzkkow9 fcdblym fg706s2 fjik90z fnj9rhh ff5xd0 f81f84o f1d0fknw f1k2sg7s f1i4xqp8 fc268be fx9k009 fxo758k fnieqtq f1t9eafa f1v1fssx f1fa3979 f1qg7kce fifjimi f8rsvsd f18fiblo f1asnnh4"
    href="https://www.bing.com"
    role="button"
    tabindex="0"
    type="button"
  >
    This is a button
  </a>
</div>
`;
